pkg/controller/controller_flow_binding_test.go

```diff
@@ -0,0 +1,195 @@
+// the bind operation completes regardless of success or failure.
+func TestServiceBindingDeleteWithAsyncBindInProgress(t *testing.T) {
+
+	for tn, state := range map[string]v2.LastOperationState{
+		"binding succeeds": v2.StateSucceeded,
+		"binding fails":    v2.StateFailed,
+	} {
+		t.Run(tn, func(t *testing.T) {
+			// Enable the AsyncBindingOperations feature
+			utilfeature.DefaultMutableFeatureGate.Set(fmt.Sprintf("%v=true", scfeatures.AsyncBindingOperations))
+			defer utilfeature.DefaultMutableFeatureGate.Set(fmt.Sprintf("%v=false", scfeatures.AsyncBindingOperations))
+
+			// GIVEN
+			ct := newControllerTest(t)
+			defer ct.TearDown()
+			ct.AsyncForBind()
+			ct.SetOSBPollBindingLastOperationReactionsState(v2.StateInProgress)
+			require.NoError(t, ct.CreateSimpleClusterServiceBroker())
+			require.NoError(t, ct.WaitForReadyBroker())
+			ct.AssertClusterServiceClassAndPlan(t)
+			assert.NoError(t, ct.CreateServiceInstance())
+			assert.NoError(t, ct.WaitForReadyInstance())
+			assert.NoError(t, ct.CreateBinding())
+			assert.NoError(t, ct.WaitForBindingInProgress())
+
+			// WHEN
+			assert.NoError(t, ct.Unbind())
+			// let's finish binding with a given state
+			ct.SetOSBPollBindingLastOperationReactionsState(state)
+
+			// THEN
+			assert.NoError(t, ct.WaitForUnbindStatus(v1beta1.ServiceBindingUnbindStatusSucceeded))
+			// at least one unbind call
+			assert.NotZero(t, ct.NumberOfOSBUnbindingCalls())
+		})
+	}
+}
+
+// TestDeleteServiceBindingRetry tests whether deletion of a service binding
+// retries after failing.
+func TestDeleteServiceBindingFailureRetry(t *testing.T) {
+	// GIVEN
+	ct := newControllerTest(t)
+	defer ct.TearDown()
+	ct.SetFirstOSBUnbindReactionsHTTPError(1, http.StatusInternalServerError)
+	require.NoError(t, ct.CreateSimpleClusterServiceBroker())
+	require.NoError(t, ct.WaitForReadyBroker())
+	ct.AssertClusterServiceClassAndPlan(t)
+	assert.NoError(t, ct.CreateServiceInstance())
+	assert.NoError(t, ct.WaitForReadyInstance())
+	assert.NoError(t, ct.CreateBinding())
+	assert.NoError(t, ct.WaitForReadyBinding())
+
+	// WHEN
+	assert.NoError(t, ct.Unbind())
+
+	// THEN
+	assert.NoError(t, ct.WaitForUnbindStatus(v1beta1.ServiceBindingUnbindStatusSucceeded))
+
+	// THEN
+	assert.NoError(t, ct.WaitForUnbindStatus(v1beta1.ServiceBindingUnbindStatusSucceeded))
+	// at least two unbind calls
+	assert.True(t, ct.NumberOfOSBUnbindingCalls() > 1)
+}
+
+// TestDeleteServiceBindingRetry tests whether deletion of a service binding
+// retries after failing an asynchronous unbind.
+func TestDeleteServiceBindingFailureRetryAsync(t *testing.T) {
+	// GIVEN
+	utilfeature.DefaultMutableFeatureGate.Set(fmt.Sprintf("%v=true", scfeatures.AsyncBindingOperations))
+	defer utilfeature.DefaultMutableFeatureGate.Set(fmt.Sprintf("%v=false", scfeatures.AsyncBindingOperations))
+
+	ct := newControllerTest(t)
+	defer ct.TearDown()
+	//async bind/unbind
+	ct.AsyncForBind()
```

Comment:
could you make those methods more readable? Right now I do not know what that method is doing without checking the implementation details.

- AsyncForInstanceProvisioning
- AsyncForInstanceUpdate
- AsyncForInstanceDeprovisioning
- AsyncForUnbind
- AsyncForBind
- SyncForBindings


e.g. `AssertOSBBasicAuth` or `SetFirstOSBProvisionReactionsHTTPError` are self-documenting and it's really easy to read the code, so please fix the methods that I've mentioned above to also be clear

Reply:
renamed